### PR TITLE
Gradient descend implementation of IVHD

### LINKED ISSUE #4: Add gradient descent version of calculating embedding
Currently simulation method is used. Use gradient descent. Treat input as parameters. It should be implemented as separate class.

Optimizer should be parametrized (e.g. SGD, Adam)

### PR DESCRIPTION
**Pull request by mjste**

### CODE REVIEW

File: ivhd/ivhd_grad.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        knn_model.fit(X)
+        _, indices = knn_model.kneighbors(X)
+        return indices[:, 1:]
+
+    def _get_remote_neighbors_indexes(self, X: np.ndarray) -> np.ndarray:
+        # for every point in X sample indices of its 'rn' remote neighbors
+        return np.random.randint(low=0, high=X.shape[0], size=(X.shape[0], self.rn))
+
+    def __sklearn_is_fitted__(self) -> bool:
+        return True  # all calculations are performed in 'transform' method
```
Comment (AdrianBesciak) [inline ivhd/ivhd_grad.py:98]: Why True is hardcoded here as a response?

**Review by AdrianBesciak — CHANGES_REQUESTED** [review]

File: ivhd/ivhd_grad.py (hunk 1 above)
Comment (mjste) [reply ivhd/ivhd_grad.py:98]: Same as in previous, non-gradient, approach - there is nothing to fit. IVHD does not have any internal parameters. You cannot fit it, and then use gained knowledge on other set. Therefore, it is always already "fitted".

**Review by AdrianBesciak — APPROVED** [review]

File: ivhd/ivhd_grad.py (hunk 1 above)
Comment (AdrianBesciak) [reply ivhd/ivhd_grad.py:98]: Thanks

**Review by BlazejNowicki — APPROVED** [review]

**Review by AdrianBesciak — APPROVED** [review]
